### PR TITLE
[lora, perf] refactor: accumulate independent MoE-LoRA gate and up projections

### PR DESCRIPTION
**Pull request by Levius-Fubuki**
## Purpose

The independent eager MoE-LoRA path currently materializes the gate and up
LoRA deltas, scales both full-width tensors, concatenates them, and launches
a separate add against the merged base projection.

This PR splits the base gate/up projection first and uses standard
`torch.addmm` to accumulate each rank-r LoRA projection directly into the
corresponding base output:

```text
before: {mm: 5, mul: 2, cat: 1, add: 1}
after:  {mm: 3, addmm: 2}
```

This removes the two full-width delta tensors, their separate scale kernels,
the concatenation, and the final residual add. The shared and fused MoE-LoRA
paths are unchanged.

The Python float scale cache is synchronized when loading the persistent
`lora_scaling` buffer. The load hook reads the source state-dict entry so it
also remains safe when the destination module is initialized on `meta`.

## Checklist Before Starting

- Searched existing issues and PRs; no duplicate gate/up eager accumulation
  change was found.
- The PR title follows VeOmni's required
  `[{modules}] {type}: {description}` format.

### Regression tests

```bash
python -m pytest -q tests/lora/test_moe_lora_eager.py \
  -k "not qwen3_5_moe"
```

The tests cover:

- the real independent Qwen3-MoE wrapper selecting two gate/up `addmm`
  accumulations;
- nonzero A/B weights and a non-unit scale;
- forward and gradient equivalence against the pre-change formula;
- scale-cache synchronization after `load_state_dict`;
- loading into a meta-initialized destination.

Result:

```text
30 passed, 6 deselected
```

The six deselected Qwen3.5 cases require the optional `fla` dependency, which
was unavailable in the benchmark environment.

### Lint and formatting

```bash
make quality
git diff --check
```

Result:

```text
ruff check: passed
ruff format --check: 589 files already formatted
git diff --check: passed
```

### RTX 4090 microbenchmark

Hardware/software: NVIDIA RTX 4090 24 GB, CUDA 13.0,
PyTorch 2.12.1+cu130. BF16 inputs, hidden size 4096,
intermediate size 14336, LoRA rank 16, 20 warmups and 100 measured iterations.

| Expert tokens | Old/New forward (ms) | Forward speedup | Old/New forward+backward (ms) | F+B speedup | Old/New forward peak allocation (MiB) |
|---:|---:|---:|---:|---:|---:|
| 8 | 0.3625 / 0.3318 | 1.09x | 3.0879 / 2.9189 | 1.06x | 243.63 / 242.75 |
| 32 | 0.4035 / 0.3205 | 1.26x | 3.0746 / 2.9460 | 1.04x | 250.63 / 246.88 |
| 128 | 0.4721 / 0.3697 | 1.28x | 3.0756 / 2.8104 | 1.09x | 277.63 / 263.63 |

The benchmark matches LoRA training semantics: hidden states and adapter
weights require gradients, while the base expert weight remains frozen.

The FP32 reference produced identical outputs. The maximum absolute gradient
difference ranged from `7.6e-6` to `1.2e-4`, reflecting the changed GEMM
accumulation order.

## API and Usage Example

No public API or configuration changes are required. The optimization applies
only to `LoraIndependentExperts` when the eager fallback is selected.

## Design & Code Changes

- Split the merged base gate/up result into views before applying LoRA.
- Accumulate gate and up LoRA projections with two `torch.addmm` calls.
- Preserve the loaded `lora_scaling` semantics by synchronizing the hot-path
  float cache from the source state dict.
- Add operator-path, numerical, gradient, state-load, and meta-init coverage.

## Limitations

These are isolated gate/up subgraph measurements from one RTX 4090 run, not
end-to-end MoE training throughput.

The benchmark used PyTorch 2.12.1+cu130, while the repository currently locks
PyTorch 2.11.0+cu130. Other GPU architectures and the Ascend eager path were
not benchmarked. Upstream CI remains the cross-platform validation gate.

## Checklist Before Submitting

- [x] Read the contribution guide
- [x] Applied the repository's locked Ruff 0.13.2 checks
- [x] Added focused regression tests
- [x] No public API or documentation update is required

## AI assistance

AI assistance was used during implementation, testing, benchmarking, review,
and PR drafting. The human contributor reviewed and takes responsibility for
the change and evidence.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved eager MoE-LoRA computations to preserve correct outputs and gradients across supported devices and data types.
  * Fixed restoration of LoRA scaling values when loading saved model state, including converted values and metadata-only device setups.

* **Tests**
  * Added coverage for output accumulation, scaling behavior, checkpoint loading, device handling, and gradient correctness.
  * Expanded validation across CPU and GPU configurations with multiple data types.

### CODE REVIEW

File: veomni/lora/moe_layers.py [hunk 1]
```diff
@@ -935,6 +935,29 @@ def __init__(
...
+            state_dict,
+            prefix,
+            local_metadata,
+            strict,
+            missing_keys,
+            unexpected_keys,
+            error_msgs,
+        )
+        if loaded_scaling is not None and loaded_scaling.numel() == 1 and not loaded_scaling.is_meta:
+            self._lora_scale_value = loaded_scaling.item()
```
[the remark this answers is not in the document]
Comment (Levius-Fubuki) [reply veomni/lora/moe_layers.py]: Fixed in 3207dd7.

The cached scale is now cast to the destination `lora_scaling` buffer dtype before extracting the Python scalar, matching the dtype conversion performed by `load_state_dict`.

The regression test loads a non-representable FP64 `1 / 3` scale into the FP32 buffer and verifies that the cache matches the restored buffer value. The numerical equivalence test now also covers CUDA FP16 and BF16 with a shared explicit `grad_outputs` tensor for forward and VJP comparison.

Validation:
- 32 passed, 6 deselected
- `make quality` passed
- `git diff --check` passed

**Comment by Levius-Fubuki** [comment]
Hi @Coach257, could you take a look when you have time?

This is a focused optimization for the independent eager MoE-LoRA gate/up path. It replaces the full-width delta/scale/cat/add chain with two `torch.addmm` accumulations.

The review feedback has been addressed, with CUDA FP16/BF16 forward and gradient equivalence coverage added. Local tests and `make quality` pass, and the RTX 4090 microbenchmark shows 1.09–1.28x forward speedup for the isolated gate/up subgraph.

**Review by Coach257 — APPROVED** [review]
